### DISCUSSION TITLE
Prompt before overwriting on first run

### DISCUSSION THREAD

**Discussion by liancea** [twpayne/chezmoi#2861] [Q&A] [answered]
Hi!

I'm currently trying to put my dotfiles under chezmoi management and noticed that by default, chezmoi will overwrite existing files during the first `apply` run that affects them, without warning. This may make sense on freshly installed environments, but on existing installations or home directories I can already see myself accidentally deleting needed files.

Is there a way to be prompted before overwriting never-before-written files like those that have been written once? I found the `--interactive` flag, but this naturally also asks about creating new files and directories, and doesn't indicate if files are going to be overwritten (even when the non-interactive counterpart would do so).

Apart from that chezmoi looks pretty nice so far, looking forward to using it!

Best regards

**Comment by twpayne** [comment, marked as answer]
Overwriting never-before-written files is exactly what you want to do when you setup your dotfiles on a new machine. This is by far the most common use case, and so needs to be the default.

The `chezmoi diff` command tells you exactly what changes `chezmoi apply` will make, and the [quick start guide](https://www.chezmoi.io/quick-start/) emphasizes that you should run `chezmoi diff` to check the changes before you run `chezmoi apply`.

Also, the quick start guide recommends using `chezmoi add` to add a file to be managed by chezmoi, then `chezmoi edit`, then `chezmoi apply`. If you follow this process, then you will get a warning if you modify the original dotfile between the `chezmoi add` and the `chezmoi apply`.

It would be possible to add a, say, `--safe` flag to `chezmoi apply` which prompts the user before modifying any existing files, but as the default has to be overwrite (for the reasons given in the first paragraph), so the `--safe` flag would have to be in the quick start documentation, and the recommendation to use `chezmoi diff` is already in the quick start documentation, so I don't think it would be a significant change to the overall user experience.

Thoughts?

**Reply by liancea** [reply]
I would not agree that losing user data is a good default in principle, i.e. when silently overwriting files on first encounter. Chezmoi simply doesn't know if it's being run in a new or existing context. That also counts for files that have been in a chezmoi-managed home for a long time, but only have been recently added to the source state, possibly at another machine (which would also elude the suggested `chezmoi` `add` -> `edit` -> `apply` workflow, even when used as recommended).

That being said, I can easily see many users wanting just this by default. Who cares about the skeleton `~/.bashrc`, for example.

I guess what confused me is that chezmoi changes its behaviour for prompting to overwrite. If it wouldn't prompt me to overwrite when the target state was modified, I'd intuitively consider `apply` dangerous and would maybe inspect the diff. Since it does, I jumped to thinking there are safeguards (which there are), making chezmoi warn me when I'm about to shoot myself in the foot. In essence, why would it warn me at all if I can just check the diff, or, from the other side, why would it not keep me from losing data unless instructed to do so (like git). This is just a personal feeling however, not an argument.

Personally, I'd love a `--safe` flag like you described. Inspecting the diff solves the problem of unintended consequences, but doesn't aid in assessing whether we, for example, replace a file because a previously applied source state was significantly changed in source (perfectly safe) or because there are some leftovers from the pre-chezmoi world (not safe). To me, the user experience would be improved a lot with `--safe`.

Then again, I agree that including a future `--safe` in the quick start documentation wouldn't improve things much. Running `chezmoi diff` basically answers all questions about what's about to happen, not only what is discussed here. If it were to come, maybe it should just be put in the reference section, like `--interactive`.

Since I didn't find anyone else bothered by this, it seems like a niche feature; I'll just take your suggestion about running `chezmoi diff` and forge that into a `--safe`-like pre-apply hook.

Thank you for your time and thank you for making chezmoi, it really is a nice tool!
